src/Symfony/Bundle/FrameworkBundle/Secrets/SodiumVault.php

```diff
@@ -0,0 +1,140 @@
+        }
+
+        $this->decryptionKey = sodium_crypto_box_keypair();
+        $this->encryptionKey = sodium_crypto_box_publickey($this->decryptionKey);
+
+        $this->export('sodium.encrypt.key', $this->encryptionKey);
+        $this->export('sodium.decrypt.key', $this->decryptionKey);
+
+        return true;
+    }
+
+    public function seal(string $name, string $value): void
+    {
+        $this->loadKeys($name);
+        $this->export($name.'.sodium', sodium_crypto_box_seal($value, $this->encryptionKey));
+    }
+
+    public function reveal(string $name): ?string
+    {
+        $this->loadKeys($name);
+
+        if (null === $this->decryptionKey || !file_exists($file = $this->pathPrefix.$name.'.sodium')) {
+            return null;
+        }
+
+        return sodium_crypto_box_seal_open(include $file, $this->decryptionKey);
+    }
+
+    public function remove(string $name): bool
+    {
+        $this->loadKeys($name);
+
+        if (!file_exists($file = $this->pathPrefix.$name.'.sodium')) {
+            return false;
+        }
+
+        return @unlink($file) || !file_exists($file);
+    }
+
+    public function list(bool $reveal = false): array
+    {
+        $this->loadKeys();
+        $secrets = [];
+        $prefix = basename($this->pathPrefix);
+        $prefixLen = \strlen($prefix);
+
+        foreach (scandir(\dirname($this->pathPrefix)) as $name) {
+            if (0 === substr_compare($name, $prefix, 0, $prefixLen) && 0 === substr_compare($name, '.sodium', -\strlen('.sodium'))) {
+                $name = substr($name, $prefixLen, -\strlen('.sodium'));
+                $secrets[$name] = null !== $this->decryptionKey && $reveal ? $this->reveal($name) : null;
+            }
+        }
+
+        return $secrets;
+    }
+
+    private function loadKeys(string $name = null): void
+    {
+        if (null !== $name && preg_match('/[^-._A-Za-z0-9\x80-\xFF]/', $name)) {
+            throw new \LogicException(sprintf('Secret name is invalid: "%s".', $name));
+        }
+
+        if (null !== $this->encryptionKey) {
+            return;
+        }
+
+        if (file_exists($this->pathPrefix.'sodium.decrypt.key')) {
+            $this->decryptionKey = include $this->pathPrefix.'sodium.decrypt.key';
+        }
+
+        if (file_exists($this->pathPrefix.'sodium.encrypt.key')) {
+            $this->encryptionKey = include $this->pathPrefix.'sodium.encrypt.key';
+        } elseif (null !== $this->decryptionKey) {
+            $this->encryptionKey = sodium_crypto_box_publickey($this->decryptionKey);
+        } else {
+            throw new \LogicException(sprintf('Encryption key not found in "%s".', $this->pathPrefix));
+        }
+    }
+
+    private function export(string $file, string $data): void
+    {
+        $name = basename($this->pathPrefix.$file);
+        $data = str_replace('%', '\x', rawurlencode($data));
+        $data = sprintf("<?php // %s on %s\n\nreturn \"%s\";\n", $name, date('r'), $data);
```

Comment:
Now dumped with some context inside, e.g.
`// dev.foo.sodium on Wed, 16 Oct 2019 17:47:31 +0200`

Reply:
Could we add that information inside the keys too? Maybe add both the human-friendly date and a parser-friendly timestamp so we can (later) provide some checker that reports if a key is "too old" and should be rotated.

Reply:
The date *is* parseable: `date('r')` is [RFC2822](http://www.faqs.org/rfcs/rfc2822.html). But I don't think we should create an API to get the date: files are made to be committed; **the git history** is the place to read their history. I don't understand what you mean with "Could we add that information inside the keys too".

Reply:
Currently a key is like this:

```php
<?php return "\x25R\x93sx\xC4\x9E\xA2X\xE7\xA7F\xF6\x22Jj\x25\xD2\x98O\xD1\x1Bd\xFD\xFD\x17\x8E\xA3\xA1iKc\xFA\xA3aa\xBCQl\xC6D\x9F\xAB\xE7E\xD6\x7Cq\x2C\x93a\xC5\xB4\xBF\xE6\x22\x97uXD\x88\xE6\xA6\x19";
```

It could be like this:

```php
<?php // key generated on Wed, 16 Oct 2019 17:47:31 +0200
return "\x25R\x93sx\xC4\x9E\xA2X\xE7\xA7F\xF6\x22Jj\x25\xD2\x98O\xD1\x1Bd\xFD\xFD\x17\x8E\xA3\xA1iKc\xFA\xA3aa\xBCQl\xC6D\x9F\xAB\xE7E\xD6\x7Cq\x2C\x93a\xC5\xB4\xBF\xE6\x22\x97uXD\x88\xE6\xA6\x19";
```

This way, a command can read that date and output messages like: *"Your keys are 6 months old. It's recommended to update them by running the "secrets:generate-keys --rotate" command.*

Reply:
Oh, I think you have an old key, please regenerate one: the very line you're commenting on is now also used for keys.